### DISCUSSION TITLE
Budget Optimizer

### DISCUSSION THREAD

**Discussion by biggielutz** [google/lightweight_mmm#189] [Q&A]
Newer to Python and having trouble with the Budget Optimizer. I run the code:

### Budget Optimization
prices = jnp.ones(mmm.n_media_channels)

n_time_periods = 10
budget = jnp.sum(media_data.mean(axis=0)) * n_time_periods
extra_features_forecast = extra_features_scaler.transform(extra_features_test)[:n_time_periods]
# run budget optimization
solution = optimize_media.find_optimal_budgets(
    n_time_periods=n_time_periods,
    media_mix_model=mmm,
    extra_features=extra_features_forecast,
    budget=budget,
    prices=prices,
    media_scaler=media_scaler,
    target_scaler=target_scaler)

budget, jnp.sum(solution.x * prices)

and get the error- 
AttributeError                            Traceback (most recent call last)
Cell In[39], line 2
      1 # both values should be almost equal
----> 2 budget, jnp.sum(solution.x * prices)

**AttributeError: 'tuple' object has no attribute 'x'**

any ideas on what I am doing wrong?

**Comment by thipokKub** [comment]
From the [source code](https://lightweight-mmm.readthedocs.io/en/latest/_modules/lightweight_mmm/optimize_media.html#find_optimal_budgets). The `optimize_media.find_optimal_budgets` return tuple of `3` items which are `solution`, `kpi_without_optim`, and `starting_values`. So I think your code should look like `solution[0].x` instead

**Reply by biggielutz** [reply]
Thanks! It hasn't worked for me yet- let me look at the source code again and try to get this right......

**Reply by biggielutz** [reply]
I used the demo https://github.com/google/lightweight_mmm/blob/main/examples/simple_end_to_end_demo.ipynb code for the optimizer and it worked perfect;y thatnks for your help....
